From 34547ea82593899a44dcb5b5d3e09de4f6e73f60 Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Mon, 20 Feb 2023 07:38:03 -0500
Subject: [PATCH 060/500] [Huawei] bpf: Use struct bpf_tcp_update_tuple as
 parameter of bpf_update_tcp_seq()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Previous patch added members in 'bpf_sock_tuple' to pass parameters to
bpf_update_tcp_seq(), which changed the size of bpf_sock_tuple.ipv4.
This make some bpf test case failed, such as 'cls_redirect'.

In order not to break the previous test cases, add a new struct
'bpf_tcp_update_tuple' to pass parameters. Also add struct
'bpf_sock_tuple' in it because ebpf prog allows only up to 5 parameters
for functions.

Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
---
 include/uapi/linux/bpf.h       | 15 ++++++++++-----
 net/core/filter.c              | 22 ++++++++++++++++------
 samples/bpf/bmc/bpf.c          | 32 +++++++++++++++++---------------
 scripts/bpf_helpers_doc.py     |  2 ++
 tools/include/uapi/linux/bpf.h | 15 ++++++++++-----
 5 files changed, 55 insertions(+), 31 deletions(-)

diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
index 8f3dd974517b..fd93710460e2 100644
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -3756,7 +3756,7 @@ union bpf_attr {
  *     Return
  *             0 on success, or a negative error in case of failure.
  *
- * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_sock_tuple *tuple, u32 len, u32 netns_id, u64 flags)
+ * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_tcp_update_tuple *tuple, u32 len, u32 netns_id, u64 flags)
  *     Description
  *             Update tcp seq, only available with kernel compiled with
  *             the **CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER** configuration
@@ -4264,10 +4264,6 @@ struct bpf_sock_tuple {
 			__be32 daddr;
 			__be16 sport;
 			__be16 dport;
-			__be32 seq;
-			__be32 rcv_len;
-			__be32 ack_seq;
-			__be32 snd_len;
 		} ipv4;
 		struct {
 			__be32 saddr[4];
@@ -4282,6 +4278,15 @@ struct bpf_xdp_sock {
 	__u32 queue_id;
 };
 
+struct bpf_tcp_update_tuple {
+	struct bpf_sock_tuple sock;
+	__u32 family;
+	__be32 seq;
+	__be32 rcv_len;
+	__be32 ack_seq;
+	__be32 snd_len;
+};
+
 #define XDP_PACKET_HEADROOM 256
 
 /* User return codes for XDP prog type.
diff --git a/net/core/filter.c b/net/core/filter.c
index 8eec132ed452..01ef91e15b24 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -6297,7 +6297,7 @@ static inline void tcp_snd_nxt_update(struct tcp_sock *tp, u32 ack_seq, u32 snd_
 }
 
 BPF_CALL_5(bpf_xdp_update_tcp_seq, struct xdp_buff *, ctx,
-	   struct bpf_sock_tuple *, tuple, u32, len, u32, netns_id, u64, flags)
+	   struct bpf_tcp_update_tuple *, tuple, u32, len, u32, netns_id, u64, flags)
 {
 	struct net *caller_net = dev_net(ctx->rxq->dev);
 	int ifindex = ctx->rxq->dev->ifindex;
@@ -6305,7 +6305,17 @@ BPF_CALL_5(bpf_xdp_update_tcp_seq, struct xdp_buff *, ctx,
 	struct tcp_sock *tp;
 	u32 seq, ack;
 
-	sk = __bpf_sk_lookup(NULL, tuple, len, caller_net,
+	if (unlikely(len != sizeof(struct bpf_tcp_update_tuple)))
+		return -1;
+
+	if (tuple->family == AF_INET)
+		len = sizeof(tuple->sock.ipv4);
+	else if (tuple->family == AF_INET6)
+		len = sizeof(tuple->sock.ipv6);
+	else
+		return -1;
+
+	sk = __bpf_sk_lookup(NULL, &tuple->sock, len, caller_net,
 			     ifindex, IPPROTO_TCP, netns_id,
 			     flags);
 	if (!sk)
@@ -6313,15 +6323,15 @@ BPF_CALL_5(bpf_xdp_update_tcp_seq, struct xdp_buff *, ctx,
 
 	tp = tcp_sk(sk);
 
-	seq = tuple->ipv4.seq;
-	ack = tuple->ipv4.ack_seq;
+	seq = tuple->seq;
+	ack = tuple->ack_seq;
 
 	if (seq == tp->copied_seq && seq == tp->rcv_nxt &&
 	    ack == tp->write_seq && ack == tp->snd_nxt) {
 		sock_owned_by_me((struct sock *)tp);
 		tp->bytes_acked = tp->bytes_sent;
-		tcp_rcv_nxt_update(tp, seq, tuple->ipv4.rcv_len);
-		tcp_snd_nxt_update(tp, ack, tuple->ipv4.snd_len);
+		tcp_rcv_nxt_update(tp, seq, tuple->rcv_len);
+		tcp_snd_nxt_update(tp, ack, tuple->snd_len);
 
 		return 0;
 	}
diff --git a/samples/bpf/bmc/bpf.c b/samples/bpf/bmc/bpf.c
index 60a8cc526c96..d1c15387c75b 100644
--- a/samples/bpf/bmc/bpf.c
+++ b/samples/bpf/bmc/bpf.c
@@ -123,21 +123,23 @@ static bool is_bmc_port(u32 port)
 	return val != NULL && *val != 0;
 }
 
-static inline void set_tcp_tuple(struct bpf_sock_tuple *tuple, char *data, u32 len)
+static inline void set_tcp_update(struct bpf_tcp_update_tuple *tuple, char *data, u32 len)
 {
 	struct iphdr *ihdr = IP_HDR(data);
 	struct tcphdr *thdr = TCP_HDR(data);
 
-	tuple->ipv4.saddr = ihdr->saddr;
-	tuple->ipv4.daddr = ihdr->daddr;
-	tuple->ipv4.sport = thdr->source;
-	tuple->ipv4.dport = thdr->dest;
-
-	tuple->ipv4.seq = __bpf_ntohl(thdr->seq);
-	tuple->ipv4.rcv_len = __bpf_ntohs(ihdr->tot_len) -
-						   sizeof(struct iphdr) - thdr->doff * 4;
-	tuple->ipv4.ack_seq = __bpf_ntohl(thdr->ack_seq);
-	tuple->ipv4.snd_len = len;
+	__builtin_memset(tuple, 0, sizeof(*tuple));
+	tuple->sock.ipv4.saddr = ihdr->saddr;
+	tuple->sock.ipv4.daddr = ihdr->daddr;
+	tuple->sock.ipv4.sport = thdr->source;
+	tuple->sock.ipv4.dport = thdr->dest;
+
+	tuple->family = AF_INET;
+	tuple->seq = __bpf_ntohl(thdr->seq);
+	tuple->rcv_len = __bpf_ntohs(ihdr->tot_len) -
+				   sizeof(struct iphdr) - thdr->doff * 4;
+	tuple->ack_seq = __bpf_ntohl(thdr->ack_seq);
+	tuple->snd_len = len;
 }
 
 /*
@@ -146,14 +148,14 @@ static inline void set_tcp_tuple(struct bpf_sock_tuple *tuple, char *data, u32 l
  */
 static int prepare_tcp(struct xdp_md *xdp, char *data, char *data_end, u32 len)
 {
-	struct bpf_sock_tuple tuple;
+	struct bpf_tcp_update_tuple tuple;
 	int old_len = data_end - data;
 	int new_len = (TCP_BODY(data) - data) + len;
 
-	set_tcp_tuple(&tuple, data, len);
+	set_tcp_update(&tuple, data, len);
 
 	if (old_len >= new_len) {
-		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple.ipv4), -1, 0))
+		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple), -1, 0))
 			return -1;
 
 		/* adjust won't fail when shrink xdp size */
@@ -162,7 +164,7 @@ static int prepare_tcp(struct xdp_md *xdp, char *data, char *data_end, u32 len)
 		if (bpf_xdp_adjust_tail(xdp, new_len - old_len))
 			return -1;
 
-		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple.ipv4), -1, 0)) {
+		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple), -1, 0)) {
 			/* shrink xdp size to rollback the buff */
 			bpf_xdp_adjust_tail(xdp, old_len - new_len);
 			return -1;
diff --git a/scripts/bpf_helpers_doc.py b/scripts/bpf_helpers_doc.py
index 31484377b8b1..f3fab6b42002 100755
--- a/scripts/bpf_helpers_doc.py
+++ b/scripts/bpf_helpers_doc.py
@@ -413,6 +413,7 @@ class PrinterHelpers(Printer):
             'struct bpf_sock_addr',
             'struct bpf_sock_ops',
             'struct bpf_sock_tuple',
+            'struct bpf_tcp_update_tuple',
             'struct bpf_spin_lock',
             'struct bpf_sysctl',
             'struct bpf_tcp_sock',
@@ -460,6 +461,7 @@ class PrinterHelpers(Printer):
             'struct bpf_sock_addr',
             'struct bpf_sock_ops',
             'struct bpf_sock_tuple',
+            'struct bpf_tcp_update_tuple',
             'struct bpf_spin_lock',
             'struct bpf_sysctl',
             'struct bpf_tcp_sock',
diff --git a/tools/include/uapi/linux/bpf.h b/tools/include/uapi/linux/bpf.h
index 639d6aafd85e..d95a9ae20c62 100644
--- a/tools/include/uapi/linux/bpf.h
+++ b/tools/include/uapi/linux/bpf.h
@@ -3756,7 +3756,7 @@ union bpf_attr {
  *     Return
  *             0 on success, or a negative error in case of failure.
  *
- * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_sock_tuple *tuple, u32 len, u32 netns_id, u64 flags)
+ * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_tcp_update_tuple *tuple, u32 len, u32 netns_id, u64 flags)
  *     Description
  *             Update tcp seq, only available with kernel compiled with
  *             the **CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER** configuration
@@ -4263,10 +4263,6 @@ struct bpf_sock_tuple {
 			__be32 daddr;
 			__be16 sport;
 			__be16 dport;
-			__be32 seq;
-			__be32 rcv_len;
-			__be32 ack_seq;
-			__be32 snd_len;
 		} ipv4;
 		struct {
 			__be32 saddr[4];
@@ -4281,6 +4277,15 @@ struct bpf_xdp_sock {
 	__u32 queue_id;
 };
 
+struct bpf_tcp_update_tuple {
+	struct bpf_sock_tuple sock;
+	__u32 family;
+	__be32 seq;
+	__be32 rcv_len;
+	__be32 ack_seq;
+	__be32 snd_len;
+};
+
 #define XDP_PACKET_HEADROOM 256
 
 /* User return codes for XDP prog type.
-- 
2.17.1

